File: scripts/fetch_chainthink.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
抓取 ChainThink 文章列表与图片，输出到 public/data/chainthink-news.json

用法：
  python scripts/fetch_chainthink.py
  python scripts/fetch_chainthink.py --limit 20 --with-detail
"""

from __future__ import annotations

import argparse
import datetime as dt
import html
import json
import pathlib
import re
import sys
import time
import urllib.error
import urllib.parse
import urllib.request
from typing import Any, Dict, List, Optional

BASE_DIR = pathlib.Path(__file__).resolve().parents[1]
OUTPUT_PATH = BASE_DIR / "public" / "data" / "chainthink-news.json"
LIST_URL = "https://chainthink.cn/zh-CN/article"
DETAIL_URL_PREFIX = "https://chainthink.cn/zh-CN/article/"
USER_AGENT = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36"
)


def http_get(url: str, timeout: int = 15) -> str:
    req = urllib.request.Request(url, headers={"User-Agent": USER_AGENT})
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        charset = resp.headers.get_content_charset() or "utf-8"
        data = resp.read()
    return data.decode(charset, errors="replace")



def decode_next_f_payload(html_text: str) -> str:
    """提取 self.__next_f.push([1,"..."]) 内层字符串并做一次反转义。"""
    marker = '[1,"'
    start = html_text.find(marker)
    if start == -1:
        raise ValueError("未找到 Next.js payload 起始位置")

    start += len(marker)
    end = html_text.rfind('"]])')
    if end == -1 or end <= start:
        raise ValueError("未找到 Next.js payload 结束位置")

    raw = html_text[start:end]
    # 只做当前页面实际需要的反转义
    return (
        raw.replace(r'\"', '"')
        .replace(r'\n', '\n')
        .replace(r'\\', r'\')
    )



def extract_json_object_after_key(text: str, key_with_quotes: str) -> Dict[str, Any]:
    idx = text.find(key_with_quotes)
    if idx == -1:
        raise ValueError(f"未找到键: {key_with_quotes}")

    after = text[idx + len(key_with_quotes):]
    obj_start = after.find('{"')
    if obj_start == -1:
        raise ValueError(f"未找到 {key_with_quotes} 对应 JSON 对象起点")

    pos = obj_start
    depth = 0
    in_string = False
    escape = False

    while pos < len(after):
        ch = after[pos]
        if in_string:
            if escape:
                escape = False
            elif ch == '\\':
                escape = True
            elif ch == '"':
                in_string = False
        else:
            if ch == '"':
                in_string = True
            elif ch in '{[':
                depth += 1
            elif ch in '}]':
                depth -= 1
                if depth == 0:
                    json_raw = after[obj_start:pos + 1]
                    return json.loads(json_raw)
        pos += 1

    raise ValueError(f"未能完整提取 {key_with_quotes} 对应 JSON 对象")



def format_publish_time(ts: Any) -> str:
    try:
        ts_int = int(ts)
    except Exception:
        return ""
    dt_obj = dt.datetime.fromtimestamp(ts_int)
    return dt_obj.strftime("%m-%d %H:%M")



def normalize_image(src: str) -> str:
    if not src:
        return ""
    src = html.unescape(str(src)).strip()
    if src.startswith("http://") or src.startswith("https://"):
        return src
    if src.startswith("/_next/image?url="):
        parsed = urllib.parse.urlparse(src)
        qs = urllib.parse.parse_qs(parsed.query)
        if qs.get("url"):
            return urllib.parse.unquote(qs["url"][0])
        return "https://chainthink.cn" + src
    if src.startswith("//"):
        return "https:" + src
    if src.startswith("/"):
        return "https://chainthink.cn" + src
    return src



def strip_tags(text: str) -> str:
    if not text:
        return ""
    text = re.sub(r"<script[\s\S]*?</script>", "", text, flags=re.I)
    text = re.sub(r"<style[\s\S]*?</style>", "", text, flags=re.I)
    text = re.sub(r"<[^>]+>", " ", text)
    text = html.unescape(text)
    text = re.sub(r"\s+", " ", text).strip()
    return text



def fetch_detail(detail_id: str) -> Dict[str, str]:
    url = DETAIL_URL_PREFIX + str(detail_id)
    try:
        page = http_get(url, timeout=15)
    except Exception:
        return {"content": "", "image": ""}

    image = ""
    m = re.search(r'"coverImage":"([^"]+)"', page)
    if m:
        image = normalize_image(m.group(1))

    content = ""
    # 优先抓 text 字段（常见为 HTML 片段或富文本）
    m = re.search(r'"text":"([\s\S]*?)","title":"', page)
    if m:
        raw_text = m.group(1)
        raw_text = raw_text.encode("utf-8").decode("unicode_escape", errors="ignore")
        content = strip_tags(raw_text)

    return {"content": content, "image": image}



def build_items(payload_text: str, limit: int, with_detail: bool) -> List[Dict[str, Any]]:
    latest = extract_json_object_after_key(payload_text, '"initialLatestData"')
    items = latest.get("list") or []
    results: List[Dict[str, Any]] = []

    for item in items[:limit]:
        info = item.get("info") or {}
        article_id = str(item.get("id") or "")
        title = str(info.get("title") or "").strip()
        if not title:
            continue
        if any(word in title for word in ("ChainThink链智库", "ChainThink", "链智库")):
            continue

        result = {
            "id": article_id,
            "title": title,
            "summary": str(info.get("abstract") or "").strip(),
            "time": format_publish_time(item.get("publishTime")),
            "published_at": item.get("publishTime"),
            "url": DETAIL_URL_PREFIX + article_id if article_id else LIST_URL,
            "tags": [tag.get("tagName") for tag in (item.get("contentTags") or []) if tag.get("tagName")],
            "isImportant": bool(item.get("isGood")),
            "coverImage": normalize_image(str(info.get("coverImage") or "")),
            "content": "",
            "source": "ChainThink",
        }

        if with_detail and article_id:
            detail = fetch_detail(article_id)
            if detail.get("image") and not result["coverImage"]:
                result["coverImage"] = detail["image"]
            if detail.get("content"):
                result["content"] = detail["content"]
            # 稍微放慢，避免太猛
            time.sleep(0.2)

        results.append(result)

    return results



def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--limit", type=int, default=20)
    parser.add_argument("--with-detail", action="store_true", default=True)
    parser.add_argument("--no-detail", action="store_true")
    args = parser.parse_args()

    with_detail = args.with_detail and not args.no_detail

    try:
        html_text = http_get(LIST_URL)
        payload = decode_next_f_payload(html_text)
        items = build_items(payload, args.limit, with_detail)

        OUTPUT_PATH.parent.mkdir(parents=True, exist_ok=True)
        output = {
            "success": True,
            "source": "python-chainthink-scraper",
            "generatedAt": dt.datetime.now().isoformat(),
            "count": len(items),
            "data": items,
        }
        OUTPUT_PATH.write_text(json.dumps(output, ensure_ascii=False, indent=2), encoding="utf-8")
        print(f"saved: {OUTPUT_PATH}")
        print(f"count: {len(items)}")
        return 0
    except urllib.error.HTTPError as e:
        print(f"HTTP error: {e.code} {e.reason}", file=sys.stderr)
        return 1
    except Exception as e:
        print(f"error: {e}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/fetch_chainthink.js
#!/usr/bin/env node

import fs from 'node:fs/promises'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const BASE_DIR = path.resolve(__dirname, '..')
const OUTPUT_PATH = path.join(BASE_DIR, 'public', 'data', 'chainthink-news.json')
const LIST_URL = 'https://chainthink.cn/zh-CN/article'
const DETAIL_URL_PREFIX = 'https://chainthink.cn/zh-CN/article/'
const USER_AGENT = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36'

function parseArgs() {
  const args = process.argv.slice(2)
  const options = { limit: 20, withDetail: true }
  for (let i = 0; i < args.length; i++) {
    const arg = args[i]
    if (arg === '--limit' && args[i + 1]) {
      options.limit = Number(args[i + 1]) || 20
      i++
    } else if (arg === '--no-detail') {
      options.withDetail = false
    } else if (arg === '--with-detail') {
      options.withDetail = true
    }
  }
  return options
}

async function httpGet(url, timeout = 15000) {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeout)
  try {
    const res = await fetch(url, {
      headers: { 'User-Agent': USER_AGENT },
      signal: controller.signal,
    })
    if (!res.ok) {
      throw new Error(`HTTP ${res.status} ${res.statusText}`)
    }
    return await res.text()
  } finally {
    clearTimeout(timer)
  }
}

function decodeNextFPayload(htmlText) {
  const pushMarker = 'self.__next_f.push([1,"'
  const start = htmlText.lastIndexOf(pushMarker)
  if (start === -1) throw new Error('未找到 Next.js payload 起始位置')
  const payloadStart = start + pushMarker.length

  // 页面尾部实际是 ...}]]\n"])</script>
  let end = htmlText.indexOf('"])</script>', payloadStart)
  if (end === -1) {
    end = htmlText.indexOf('"])', payloadStart)
  }
  if (end === -1 || end <= payloadStart) throw new Error('未找到 Next.js payload 结束位置')

  const raw = htmlText.slice(payloadStart, end)
  return raw
    .replace(/\\"/g, '"')
    .replace(/\\n/g, '\n')
    .replace(/\\\\/g, '\\')
}

function extractJsonObjectAfterKey(text, keyWithQuotes) {
  const idx = text.indexOf(keyWithQuotes)
  if (idx === -1) throw new Error(`未找到键: ${keyWithQuotes}`)
  const after = text.slice(idx + keyWithQuotes.length)
  const objStart = after.indexOf('{"')
  if (objStart === -1) throw new Error(`未找到 ${keyWithQuotes} 对应 JSON 对象起点`)

  let pos = objStart
  let depth = 0
  let inString = false
  let escape = false

  while (pos < after.length) {
    const ch = after[pos]
    if (inString) {
      if (escape) {
        escape = false
      } else if (ch === '\\') {
        escape = true
      } else if (ch === '"') {
        inString = false
      }
    } else {
      if (ch === '"') {
        inString = true
      } else if (ch === '{' || ch === '[') {
        depth++
      } else if (ch === '}' || ch === ']') {
        depth--
        if (depth === 0) {
          const jsonRaw = after.slice(objStart, pos + 1)
          return JSON.parse(jsonRaw)
        }
      }
    }
    pos++
  }

  throw new Error(`未能完整提取 ${keyWithQuotes} 对应 JSON 对象`)
}

function formatPublishTime(ts) {
  const n = Number(ts)
  if (!Number.isFinite(n)) return ''
  const d = new Date(n * 1000)
  const mm = String(d.getMonth() + 1).padStart(2, '0')
  const dd = String(d.getDate()).padStart(2, '0')
  const hh = String(d.getHours()).padStart(2, '0')
  const mi = String(d.getMinutes()).padStart(2, '0')
  return `${mm}-${dd} ${hh}:${mi}`
}

function normalizeImage(src) {
  if (!src) return ''
  const value = String(src).trim()
  if (value.startsWith('http://') || value.startsWith('https://')) return value
  if (value.startsWith('/_next/image?url=')) {
    try {
      const parsed = new URL(value, 'https://chainthink.cn')
      const rawUrl = parsed.searchParams.get('url')
      return rawUrl ? decodeURIComponent(rawUrl) : parsed.toString()
    } catch {
      return `https://chainthink.cn${value}`
    }
  }
  if (value.startsWith('//')) return `https:${value}`
  if (value.startsWith('/')) return `https://chainthink.cn${value}`
  return value
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

function stripTags(text) {
  return String(text || '')
    .replace(/<script[\s\S]*?<\/script>/gi, '')
    .replace(/<style[\s\S]*?<\/style>/gi, '')
    .replace(/<[^>]+>/g, ' ')
    .replace(/&nbsp;/g, ' ')
    .replace(/&amp;/g, '&')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/\s+/g, ' ')
    .trim()
}

async function fetchDetail(detailId) {
  if (!detailId) return { content: '', image: '' }
  try {
    const page = await httpGet(`${DETAIL_URL_PREFIX}${detailId}`)
    const imageMatch = page.match(/"coverImage":"([^"]+)"/)
    const textMatch = page.match(/"text":"([\s\S]*?)","title":"/)
    let content = ''
    if (textMatch?.[1]) {
      const rawText = textMatch[1]
        .replace(/\\n/g, '\n')
        .replace(/\\"/g, '"')
        .replace(/\\\\/g, '\\')
      content = stripTags(rawText)
    }
    return {
      content,
      image: normalizeImage(imageMatch?.[1] || ''),
    }
  } catch {
    return { content: '', image: '' }
  }
}

async function buildItems(payloadText, limit, withDetail) {
  const latest = extractJsonObjectAfterKey(payloadText, '"initialLatestData"')
  const items = Array.isArray(latest.list) ? latest.list : []
  const results = []

  for (const item of items.slice(0, limit)) {
    const info = item?.info || {}
    const articleId = String(item?.id || '')
    const title = String(info?.title || '').trim()
    if (!title) continue
    if (['ChainThink链智库', 'ChainThink', '链智库'].some(word => title.includes(word))) continue

    const result = {
      id: articleId,
      title,
      summary: String(info?.abstract || '').trim(),
      time: formatPublishTime(item?.publishTime),
      published_at: item?.publishTime || null,
      url: articleId ? `${DETAIL_URL_PREFIX}${articleId}` : LIST_URL,
      tags: Array.isArray(item?.contentTags) ? item.contentTags.map(tag => tag?.tagName).filter(Boolean) : [],
      isImportant: Boolean(item?.isGood),
      coverImage: normalizeImage(info?.coverImage || ''),
      content: '',
      source: 'ChainThink',
    }

    if (withDetail && articleId) {
      const detail = await fetchDetail(articleId)
      if (detail.image && !result.coverImage) result.coverImage = detail.image
      if (detail.content) result.content = detail.content
      await sleep(200)
    }

    results.push(result)
  }

  return results
}

async function main() {
  const { limit, withDetail } = parseArgs()
  const htmlText = await httpGet(LIST_URL)
  const payload = decodeNextFPayload(htmlText)
  const items = await buildItems(payload, limit, withDetail)

  await fs.mkdir(path.dirname(OUTPUT_PATH), { recursive: true })
  const output = {
    success: true,
    source: 'node-chainthink-scraper',
    generatedAt: new Date().toISOString(),
    count: items.length,
    data: items,
  }
  await fs.writeFile(OUTPUT_PATH, JSON.stringify(output, null, 2), 'utf8')
  console.log(`saved: ${OUTPUT_PATH}`)
  console.log(`count: ${items.length}`)
}

main().catch(err => {
  console.error(`error: ${err.message}`)
  process.exit(1)
})
